src/adc.rs

```diff
@@ -0,0 +1,858 @@
+                inserted_channel: None,
+                default_sample_time: SampleTime::Point_55_5,
+            }
+        }
+    }
+}
+
+/// Enabled ADC (type state)
+pub struct Enabled;
+/// Disabled ADC (type state)
+pub struct Disabled;
+
+/// Type state trait for a enabled ADC channel
+pub trait ED {}
+/// Enabled ADC (type state)
+impl ED for Enabled {}
+/// Disabled ADC (type state)
+impl ED for Disabled {}
+
+adc_pins!(ADC0,
+    PA0<Analog> => 0,
+    PA1<Analog> => 1,
+    PA2<Analog> => 2,
+    PA3<Analog> => 3,
+    PA4<Analog> => 4,
+    PA5<Analog> => 5,
+    PA6<Analog> => 6,
+    PA7<Analog> => 7,
+    PB0<Analog> => 8,
+    PB1<Analog> => 9,
+    PC0<Analog> => 10,
+    PC1<Analog> => 11,
+    PC2<Analog> => 12,
+    PC3<Analog> => 13,
+    PC4<Analog> => 14,
+    PC5<Analog> => 15,
+    Temperature<Enabled> => 16,
+    Vrefint<Enabled> => 17,
+);
+
+adc_pins!(ADC1,
+    PA0<Analog> => 0,
+    PA1<Analog> => 1,
+    PA2<Analog> => 2,
+    PA3<Analog> => 3,
+    PA4<Analog> => 4,
+    PA5<Analog> => 5,
+    PA6<Analog> => 6,
+    PA7<Analog> => 7,
+    PB0<Analog> => 8,
+    PB1<Analog> => 9,
+    PC0<Analog> => 10,
+    PC1<Analog> => 11,
+    PC2<Analog> => 12,
+    PC3<Analog> => 13,
+    PC4<Analog> => 14,
+    PC5<Analog> => 15,
+);
+
+/// Analog to Digital Converter
+pub struct Adc<ADC, ED> {
+    rb: ADC,
+    config: config::AdcConfig,
+    _enabled: PhantomData<ED>,
+}
+
+// applies to all ADCs
+impl<ADC> Adc<ADC, Disabled> {
+    /// Set ADC clock div
+    pub fn set_clock(&mut self, clock: config::Clock, rcu: &mut Rcu) {
+        use self::config::Clock::*;
+
+        match clock {
+            Apb2_div_2 | Apb2_div_4 | Apb2_div_6 | Apb2_div_8 => unsafe {
+                rcu.regs
+                    .cfg0
+                    .modify(|_, w| w.adcpsc_1_0().bits(clock as u8).adcpsc_2().clear_bit());
+            },
+            Apb2_div_12 | Apb2_div_16 => unsafe {
+                rcu.regs.cfg0.modify(|_, w| {
+                    w.adcpsc_1_0()
+                        .bits(clock as u8 >> 2)
+                        .adcpsc_2()
+                        .bit(clock as u8 & 0x1 == 0x1)
+                });
+            },
+        }
+    }
+
+    /// Applies all fields in AdcConfig
+    pub fn apply_config(&mut self, config: config::AdcConfig) {
+        self.config = config;
+    }
+
+    /// Sets the sampling resolution
+    pub fn set_resolution(&mut self, resolution: config::Resolution) {
+        self.config.resolution = resolution;
+    }
+
+    /// Sets the DR register alignment to left or right
+    pub fn set_align(&mut self, align: config::Align) {
+        self.config.align = align;
+    }
+
+    /// Enables and disables scan mode
+    pub fn set_scan(&mut self, scan: config::Scan) {
+        self.config.scan = scan;
+    }
+
+    /// Enables and disables continuous mode
+    pub fn set_continuous(&mut self, continuous: config::Continuous) {
+        self.config.continuous = continuous;
+    }
```

Comment:
Does this methods are necessary ? Because they are already implemented in AdcConfig ?

Reply:
https://docs.rs/stm32f4xx-hal/0.8.3/stm32f4xx_hal/adc/struct.Adc.html#method.set_clock